### PR TITLE
Implemented merging from other resource files

### PR DESCRIPTION
**Pull request by R4H33M**
I implemented a TODO feature - being able to merge from other resource files.
This fixes one of the issues mentioned on #7.
Also, I changed it from "Merge With" to "Merge From" as the two files being merged are not being combined.
What do you think of the implementation?

### CODE REVIEW

**Comment by scottmc** [comment]
I built it and trying opening ResourceEdit.rsrc with it, but it crashes on my Haiku.  I'm running gcc2/gcc8 hybrid here.

**Comment by scottmc** [comment]
ResourceEdit: ../haiku-git/src/system/libroot/posix/malloc_hoard2/superblock.h:171:class BPrivate::block * BPrivate::superblock::getBlock(): getNumAvailable() == 0
Kill Thread

File: source/MainWindow.cpp [hunk 1]
```diff
@@ -224,6 +224,9 @@ BMessage(B_ABOUT_REQUESTED));
 	fResourceList->AddStatusView(fStatsBox);
 	fStatsBox->AddChild(fStatsString);
 
+	fMergePanel = new BFilePanel(B_OPEN_PANEL, new BMessenger(this), NULL, 0, true,
```
Comment (korli) [inline source/MainWindow.cpp]: BFilePanel constructor will copy the BMessenger and the BMessage. new shouldn't be used.

Comment (R4H33M) [reply source/MainWindow.cpp]: I tried your suggestion, but it doesn't work without "new", because BFilePanel requires pointers in it's constructor.

**Comment by R4H33M** [comment]
> ResourceEdit: ../haiku-git/src/system/libroot/posix/malloc_hoard2/superblock.h:171:class BPrivate::block * BPrivate::superblock::getBlock(): getNumAvailable() == 0
> Kill Thread

From what I see, this issue is prevalent on current, pre-PR ResourceEdit as well, so it is not being caused by my additions. Please try building and testing HaikuArchives/ResourceEdit, and you should see this same crash - (it's random so might take some tries). It seems to be something with to do with memory management... Should I create a ticket? 

As the crash doesn't happen consistently (again, from what I am experiencing on my Haiku), can you still test and merge my PR, after you confirm that it is unrelated to this issue?

File: source/MainWindow.cpp (hunk 1 above)
Comment (owenca) [reply source/MainWindow.cpp]: Requiring pointers doesn’t mean requiring `new`. You can pass the address/reference of a `BMessenger` object. If the `BFilePanel` constructor doesn’t take the ownership of the object, the caller is responsible for destroying it.

**Comment by scottmc** [comment]
I'd switch the "Merge From", back to "Merge With".  Also I compiled the current master and wasn't able to trigger the error, but was getting it about half the time with the merge code added.  I only tried it a few times, but wasn't able to get it to crash.  I'll try it again later after the new issue it handledcas discussed by Owen and Korli.

File: source/MainWindow.cpp (hunk 1 above)
Comment (R4H33M) [reply source/MainWindow.cpp]: I've now researched addresses and pointers, but I don't find anything that works. I also searched online for uses of BFilePanel, and they all use the same implementation with the new keyword.

The solution I have been able to come up with is initialising the variable and storing it beforehand - (without new), and then pass a reference? Is this what I should do?

Comment (owenca) [reply source/MainWindow.cpp]: Yes, and [here](https://github.com/HaikuArchives/Filer/blob/master/sources/AutoFilerTab.cpp#L66) is an example.

**Comment by R4H33M** [comment]
I've switched it back to "Merge With", and I have removed the use of the "new" keyword.

File: source/MainWindow.cpp [hunk 2]
```diff
@@ -224,6 +224,12 @@ BMessage(B_ABOUT_REQUESTED));
 	fResourceList->AddStatusView(fStatsBox);
 	fStatsBox->AddChild(fStatsString);
 
+	BMessenger panelMessenger(this);
+	BMessage panelMessage(MSG_MERGE_OPEN_DONE);
+
+	fMergePanel = new BFilePanel(B_OPEN_PANEL, &panelMessenger, NULL, 0, true,
```
Comment (owenca) [inline source/MainWindow.cpp]: You need to delete `fMergePanel` in the destructor.

File: source/MainWindow.cpp [hunk 3]
```diff
@@ -386,10 +392,20 @@ MainWindow::MessageReceived(BMessage* msg)
...
+			fMergePanel->Show();
+			break;
+
+		case MSG_MERGE_OPEN_DONE:
+		{
+			//load the selected file
+			entry_ref mergeRef;
+			while (fMergePanel->GetNextSelectedRef(&mergeRef) == B_OK) {
+				fAssocEntry = new BEntry(&mergeRef, true);
+				_Load();
```
Comment (owenca) [inline source/MainWindow.cpp]:
```suggestion
				_Load();
				delete fAssocEntry;
```

File: source/MainWindow.cpp [hunk 4]
```diff
@@ -386,10 +392,20 @@ MainWindow::MessageReceived(BMessage* msg)
...
+				fUnsavedChanges = true;
+			}
 			break;
```
Comment (owenca) [inline source/MainWindow.cpp]:
```suggestion
			fAssocEntry = NULL;
				// or reset it to its previous value,
				// which must be saved before the while-loop.
			break;
```

**Comment by R4H33M** [comment]
Improved code safety as requested.

File: source/MainWindow.cpp [hunk 5]
```diff
@@ -386,10 +393,21 @@ MainWindow::MessageReceived(BMessage* msg)
...
+			break;
+
+		case MSG_MERGE_OPEN_DONE:
+		{
+			//load the selected file
+			entry_ref mergeRef;
+			while (fMergePanel->GetNextSelectedRef(&mergeRef) == B_OK) {
+				fAssocEntry = new BEntry(&mergeRef, true);
+				_Load();
+				fUnsavedChanges = true;
```
Comment (owenca) [inline source/MainWindow.cpp]: You missed `delete fAssocEntry;`.

File: source/MainWindow.cpp [hunk 6]
```diff
@@ -386,10 +393,21 @@ MainWindow::MessageReceived(BMessage* msg)
...
+				fAssocEntry = NULL;
```
Comment (owenca) [inline source/MainWindow.cpp]: This should be outside the `while` block as I suggested. You need to make sure that setting it to `NULL` is the right thing to do. As my comments in the suggested code explained, you might have to save `fAssocEntry` before the loop and restore it afterwards.

Comment (R4H33M) [reply source/MainWindow.cpp]: Since it gets used for saving the file, I will save it before the loop and restore it afterwards.

**Comment by R4H33M** [comment]
I made the requested changes, and also fixed a small typo I discovered.
Is the implementation safe? Still getting my head around pointers, but I think this is the best way to do it.

File: source/Constants.h [hunk 7]
```diff
@@ -44,6 +44,8 @@
 #define MSG_SETTINGS_REVERT	'm023'
 #define MSG_SETTINGS_CLOSED	'm024'
 
+#define MSG_MERGE_OPEN_DONE	'm025'
```
Comment (owenca) [inline source/Constants.h]: You don’t need it. See below.

File: source/MainWindow.cpp [hunk 8]
```diff
@@ -224,6 +224,12 @@ BMessage(B_ABOUT_REQUESTED));
 	fResourceList->AddStatusView(fStatsBox);
 	fStatsBox->AddChild(fStatsString);
 
+	BMessenger panelMessenger(this);
+	BMessage panelMessage(MSG_MERGE_OPEN_DONE);
```
Comment (owenca) [inline source/MainWindow.cpp]: You don’t need it.

File: source/MainWindow.cpp (hunk 2 above)
Comment (owenca) [inline source/MainWindow.cpp]:
```suggestion
	fMergePanel = new BFilePanel(B_OPEN_PANEL, &panelMessenger);
```

File: source/MainWindow.cpp [hunk 9]
```diff
@@ -224,6 +224,12 @@ BMessage(B_ABOUT_REQUESTED));
...
+
+	fMergePanel = new BFilePanel(B_OPEN_PANEL, &panelMessenger, NULL, 0, true,
+		&panelMessage);
```
Comment (owenca) [inline source/MainWindow.cpp]: Delete this line.

File: source/MainWindow.cpp [hunk 10]
```diff
@@ -386,10 +393,22 @@ MainWindow::MessageReceived(BMessage* msg)
 			break;
 
 		case MSG_MERGEWITH:
-			PRINT(("[MSG_MERGEWITH]: Not yet implemented."));
-			// TODO: Implement.
-			// "Merge from..." might be a better idea actually.
+			fMergePanel->Show();
+			break;
+
+		case MSG_MERGE_OPEN_DONE:
```
Comment (owenca) [inline source/MainWindow.cpp]:
```suggestion
		case B_REFS_RECEIVED:
```

**Comment by R4H33M** [comment]
> What happens if both resource and non-resource files are selected in the merge file panel? You probably need to add file type filtering.
> 
> The default Open button should be relabeled to Merge.

This filter could also be used by the main Open panel right? Right now ResourceEdit open all files, whether they are valid resource files or not.

**Comment by owenca** [comment]
Yes, the same filter can be used there.

**Comment by owenca** [comment]
Also:
- Change “Open” in the window title to “Merge”.
- The Merge button should be enabled only if two or more resource files are selected.

**Comment by R4H33M** [comment]
I thought that the user would also be able to select just one file to be merged as it would be merged with the file that they already had opened.
For example, I open up X.rscrs and I merge it with Y.rscrs, getting both of them merged in (unsaved) X.rscs. Isn't this how it is supposed to be?

**Comment by owenca** [comment]
I haven’t used the application, but if you can select just one file to merge with an already opened file, then disregard the second bullet above. :)

**Comment by scottmc** [comment]
This application lets you see the resources of a binary file, this also includes executables, so ResourceEdit can open itself for example.  But you probably don't want to allow editing of execuatles.  So I do'nt know that we want to filter out what a user is allowed to open, but we may want to filter out what they are allowed to write to.  This would be outside the scope of this issue.  This issue is just to enable the merge from menu option.  You might keep track of all these types of issues you think of and open new issues here on github for them.  There are times where you would want to open the resources from an application, but then "Export" them to an rdef files for example.  This is another feature that isn't even listed on the menu, but it is some other resource editors.  I think quickres has that option.
So for now I'd leave out the filtering/blocking, perhaps open a new issue to add it and attach any code you have written for it there.  we may need to discuss the issue further there before adding it or not.

**Comment by R4H33M** [comment]
I've created an issue: #19
I have also implemented the changes @owenca requested.
Ready for merge?

File: source/MainWindow.cpp [hunk 11]
```diff
@@ -224,6 +224,12 @@ BMessage(B_ABOUT_REQUESTED));
 	fResourceList->AddStatusView(fStatsBox);
 	fStatsBox->AddChild(fStatsString);
 
+	BMessenger panelMessenger(this);
+
+	fMergePanel = new BFilePanel(B_OPEN_PANEL, &panelMessenger);
+	fMergePanel->SetButtonLabel(B_DEFAULT_BUTTON, "Merge");
+	fMergePanel->SetRefFilter();
```
Comment (owenca) [inline source/MainWindow.cpp]: Does this even compile? Anyway, you can take it out.

This is a good place to change the default window title to "ResourceEdit: Merge".

**Comment by R4H33M** [comment]
Oops, I forgot about that one. Alright, I've removed it now.

**Comment by owenca** [comment]
Can you post a screenshot of the merge panel?

**Comment by R4H33M** [comment]
image

The file open panel, right?

**Pull request merged**